components/dashboard/src/custom-hooks/usePagination.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { useMemo } from "react";
+
+export const usePagination = (props: {
```

Comment:
I'm not sure why this needs to be a hook. AIUI this returns an array of page numbers based on its arguments. I think the `calculatedPagination` function could be inlined into the `Pagination` component and called there directly.

Reply:
The main reason in extracting it was for readability and ease of maintenance. The usePagination logic is quite complex and it would be much harder to maintain with other noise around it.

Reply:
completely agree to make it a separate function rather than doing it inline in the component. I just wonder if it can be an ordinary function rather than a hook.

Reply:
Aha. I think technically yes, but I did want to make use of the `useMemo` for better performance, and therefore it has to be either a Hook or a React component. Do you think it would be overhead to maintain?

Reply:
I'm thinking of converting the hook to just a function called `getPageNumbers` (or whatever), that takes the same arguments as  the hook does now and returns an array of page numbers (as the hook does now).

If the function is created outside the `Pagination` component, it won't need to be memoized because its identity won't change between renders of the `Pagination` component.

Reply:
Hmm, makes sense. The new function will then have to be called inside a useEffect, right? Whenever each of the args change. Also let me know if a quick sync would be faster 😬

Reply:
I think there is no need for a `useEffect` - the function is just a helper for turning current `props` into an array of page numbers - no synchronization with anything external that would need a `useEffect`.

essentially the pattern is:

```ts
const func getPageNumbersFromProps(...) => { return [1,2,3,4] }

const MyComponent = (props) => {
  const xs = getPageNumbersFromProps(props)

  return (<ul>
    xs.map(x => <li>x</li>
  </ul>)
}
```

Reply:
Ah sorry not sure why I said useEffect I meant useState for updating the values, but they do get updated in the parent component. Will give this a shot after lunch.